### PR TITLE
Improve docs for `std.typecons.scoped`

### PR DESCRIPTION
**Pull request by ntrel**
Show automatic destruction of `scoped` result.
Show how `scoped` can lead to invalid class references.
Show scoped member variable initialization.
Move alias example before *Restrictions* section, which uses alias.

### CODE REVIEW

File: std/typecons.d [hunk 1]
```diff
@@ -5141,7 +5141,15 @@ therefore avoiding the overhead of $(D new). This facility is unsafe;
 it is the responsibility of the user to not escape a reference to the
 object outside the scope.
 
-Note: it's illegal to move a class reference even if you are sure there
+The class's destructor will be called when the result of `scoped()` is 
+itself destroyed.
+
+Class instances can be embedded in a `class` or `struct` as member
```
Comment (kuettler) [inline std/typecons.d]: This sentence feels like a random fact. I see two points here: (a) With scoped class instance can be placed like struct instances and (b) the type `typeof(scoped!Class(args))` has to be used to use a scoped class as a member variable. The second point is actual information (I did not know that before). Maybe you can make this point more pronounced. In addition a one-sentence explanation would be nice.

Comment (ntrel) [inline std/typecons.d]: [Updated](https://github.com/D-Programming-Language/phobos/pull/3016/files#diff-4e008aedb3026d4a84f58323e53bf017R5147)

Please suggest any additional wording if needed.

**Comment by ntrel** [comment]
ping

File: std/typecons.d [hunk 2]
```diff
@@ -5141,7 +5141,16 @@ therefore avoiding the overhead of $(D new). This facility is unsafe;
 it is the responsibility of the user to not escape a reference to the
 object outside the scope.
 
-Note: it's illegal to move a class reference even if you are sure there
+The class's destructor will be called when the result of `scoped()` is
```
Comment (schveiguy) [inline std/typecons.d]: possessive for an 's' ending word is just a trailing apostrophe: `class' destructor`

**Comment by schveiguy** [comment]
Other than the nitpick, looks good.

File: std/typecons.d (hunk 2 above)
Comment (ntrel) [inline std/typecons.d]: Fixed, thanks.

File: std/typecons.d [hunk 3]
```diff
@@ -5141,7 +5141,16 @@ therefore avoiding the overhead of $(D new). This facility is unsafe;
...
+The class destructor will be called when the result of `scoped()` is 
+itself destroyed.
+
+Scoped class instances can be embedded in a parent `class` or `struct`, 
+just like a child struct instance. Scoped member variables must have 
+type `typeof(scoped!Class(args))`, and be initialized with a call to 
+scoped. See below for an example.
+
+Note:
+It's illegal to move a class reference even if you are sure there
```
Comment (quickfur) [inline std/typecons.d]: What does it mean to "move a class reference"? Assign it to another class reference? Call `std.algorithm.mutation.move` on it?

Comment (ntrel) [inline std/typecons.d]: Yes, I didn't like this language either, but I didn't know how to improve it, so I left it as-is.

Comment (quickfur) [inline std/typecons.d]: But I actually don't understand what it means. What *does* it mean "it's illegal to move a class reference"? What does "move" mean in this context?

Comment (ntrel) [inline std/typecons.d]: I'm not sure, I didn't write it.

Comment (quickfur) [inline std/typecons.d]: Alright, let's leave this to a future PR. Better to have a partially-better doc than to have no improvements at all.

File: std/typecons.d [hunk 4]
```diff
@@ -5205,34 +5215,68 @@ unittest
...
-
-    // Use with alias
-    alias makeScopedA = scoped!A;
-    auto a6 = makeScopedA();
-    auto a7 = makeScopedA();
+    // Stack-allocate
+    auto b1 = B(5);
+    aRef = b1.a;
+    assert(aRef.x == 5);
+    // Heap-allocate
```
Comment (quickfur) [inline std/typecons.d]: Nitpick: should have blank line above the comment, for better readability.

File: std/typecons.d (hunk 3 above)
Comment (schveiguy) [inline std/typecons.d]: It's a poorly written comment. Should be "It's illegal to move an object's data". It's perfectly legal to move a pointer/reference.

Comment (quickfur) [inline std/typecons.d]: I still don't quite understand exactly what that means, though. How do you "move an object's data" if said object is being held by `Scoped`? Copy it elsewhere and invalidate the scoped reference? Manually overwrite the scoped reference pointer? I'm not seeing the point of that comment. Obviously, I'm missing something obvious, but that just shows how badly written that comment is? :-P

**Comment by ntrel** [comment]
Added newlines, `destroy(b1)` and simple *Params:* section for `scoped(args)`.
Fixed note about moving a class instance - see ffc9ae4.

**Comment by ntrel** [comment]
ping

**Comment by DmitryOlshansky** [comment]
LGTM

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled off

**Comment by DmitryOlshansky** [comment]
Fails auto-tester somewhere deeper in std.typcons, memory corruption?

**Comment by quickfur** [comment]
ping @ntrel

**Comment by quickfur** [comment]
The tester is complaining about trailing whitespace. This should be a trivial fix. Anybody?

File: std/typecons.d [hunk 5]
```diff
@@ -5205,34 +5217,76 @@ unittest
...
+        A invalid = scoped!A();
+    }
 
+    // Use with alias
+    alias makeScopedA = scoped!A;
+    auto a3 = makeScopedA();
+    auto a4 = makeScopedA(1);
+    
     // Restrictions
     static assert(!is(typeof({
```
Comment (quickfur) [inline std/typecons.d]: This assert is failing for some reason. (Tested on git HEAD, I had to rebase on Phobos master to get it to compile, but `make -f posix.mak unittest` aborts with an assertion failure on this line.)

Comment (quickfur) [inline std/typecons.d]: I'm guessing something broke with `scoped`, and it now compiles with `std.algorithm.move` even though it should not.

**Comment by DmitryOlshansky** [comment]
Any plans to revive this @ntrel ?

**Comment by ntrel** [comment]
No plans really, I've not been doing any development for a while. Thanks for reviews though.

**Comment by DmitryOlshansky** [comment]
> I've not been doing any development for a while. Thanks for reviews though.

Then I'm going to close it, feel free to create a new pull.

File: std/typecons.d (hunk 5 above)
Comment (ntrel) [inline std/typecons.d]: I think `move` probably always did work with `scoped`. In my updated https://github.com/D-Programming-Language/phobos/pull/4129, I changed the assert to `version(Bug)` instead.

**Pull request closed without merging**